src/webgpu/api/validation/image_copy/image_copy.ts

```diff
@@ -98,6 +98,72 @@ export class ImageCopyTest extends ValidationTest {
       usage: GPUTextureUsage.COPY_SRC | GPUTextureUsage.COPY_DST,
     });
   }
+
+  testBuffer(
+    buffer: GPUBuffer,
+    texture: GPUTexture,
+    textureDataLayout: GPUImageDataLayout,
+    size: GPUExtent3D,
+    {
+      method,
+      dataSize,
+      success,
+      submit = false,
+    }: {
+      method: ImageCopyType;
+      dataSize: number;
+      success: boolean;
+      /** If submit is true, the validaton error is expected to come from the submit and encoding
+       * should succeed. */
+      submit?: boolean;
+    }
+  ): void {
+    switch (method) {
+      case 'WriteTexture': {
+        const data = new Uint8Array(dataSize);
+
+        this.expectValidationError(() => {
+          this.device.queue.writeTexture({ texture }, data, textureDataLayout, size);
+        }, !success);
+
+        break;
+      }
+      case 'CopyB2T': {
+        const encoder = this.device.createCommandEncoder();
+        encoder.copyBufferToTexture({ buffer, ...textureDataLayout }, { texture }, size);
+
+        if (submit) {
+          const cmd = encoder.finish();
+          this.expectValidationError(() => {
+            this.device.queue.submit([cmd]);
+          }, !success);
+        } else {
+          this.expectValidationError(() => {
+            encoder.finish();
+          }, !success);
+        }
```

Comment:
You can use `createEncoder` + `validateFinishAndSubmit` to simplify these a bit.

Reply:
I don't understand why `validateFinishAndSubmit` needs an extra argument? It basically does two things, finish and submit, and it takes two boolean arguments for whether each of them succeed. `validateFinishAndSubmitGivenState` may be too specific for this use-case, but it's just a wrapper around `validateFinishAndSubmit` for convenience so you don't have to use it.

As for why the default argument didn't work, it's because you would have to add it to the function definition (in the constructor of CommandBufferMaker), not the type definition of the field. (That class is set up really weirdly so that destructuring its methods works, so it's confusing.)

Reply:
i.e.:
```ts
const { encoder, validateFinishAndSubmit } = this.createEncoder('non-pass');
encoder.copyBufferToTexture({ buffer, ...textureDataLayout }, textureCopyView, size);
validateFinishAndSubmit(success, submit);
```

Reply:
Oops. I don't remember why I insisted on using validateFinishAndSubmitGivenState and didn't try to use validateFinishAndSubmit directly. Your suggestion is correct. Thanks!